SourceCode/h14/optimizations_02.cpp

```diff
@@ -0,0 +1,1731 @@
+                    assert(!uvValues.IsNull() && uvValues.IsArray());
+                    loadUVs(uvValues.GetArray());
+                }
+
+                const rapidjson::Value& trianglesValue 
+                            = itr->FindMember("triangles")->value;
+                assert(!trianglesValue.IsNull() && trianglesValue.IsArray());
+                loadTriangles(trianglesValue.GetArray());
+
+                normalizeVertexNormals();
+            }
+        }
+    }
+
+    std::string getFileName() const {
+        return fileName;
+    }
+    unsigned int getImageWidth() const {
+        return settings.imageSettings.width;
+    }
+    unsigned int getImageHeight() const {
+        return settings.imageSettings.height;
+    }
+    Vector& getBgColor() {
+        return settings.bgColor;
+    }
+    unsigned int getBucketSize() const {
+        return settings.imageSettings.bucketSize;
+    }
+    std::vector<Mesh>& getMeshes() {
+        return meshes;
+    }
+    Vector& getCameraPosition() {
+        return cameraPosition;
+    }
+    Matrix& getInitialTransformation() {
+        return initialTransformation;
+    }
+    std::vector<Light>& getLights() {
+        return lights;
+    }
+    std::vector<Material>& getMaterials() {
+        return materials;
+    }
+    std::map<std::string, std::shared_ptr<Texture>>& getTextures() {
+        return textures;
+    }
+
+    void setFileName(const std::string fileName) {
+        this->fileName = fileName;
+    }
+};
+
+constexpr float rayStartOffset = 0.3;
+constexpr float rayStartOffsetReflective = 0.3;
+constexpr float rayStartOffsetRefractive = 0.3;
+constexpr unsigned int maxRayDepth = 6;
+
+struct AABB {
+    Vector min {std::numeric_limits<float>::max()};
+    Vector max {std::numeric_limits<float>::lowest()};
+
+    Vector& get(unsigned int index) {
+        if(index == 0) return min;
+        else return max;
+    }
+};
+
+struct MeshToTriangle {
+    std::shared_ptr<Mesh> mesh;
+    std::vector<std::shared_ptr<Triangle>> triangles;
+};
+
+class KDNode {
+    int children[2] = {-1, -1};
+    struct AABB aabb;
+    std::vector<struct MeshToTriangle> intersected;
+public:
+    KDNode() = default;
+
+    void addValues(AABB& aabb, std::vector<struct MeshToTriangle>& intersected) {
+        this->aabb = aabb;
+        this->intersected = intersected;
+    }
+
+    void setChildren(const int index0, const int index1) {
+        children[0] = index0;
+        children[1] = index1;
+    }
+    void setAABB(struct AABB& aabb) {
+        this->aabb = aabb;
+    }
+    void setIntersected(const std::vector<struct MeshToTriangle>& intersected) {
+        this->intersected = intersected;
+    }
+
+    int* getChildren() {
+        return children;
+    }
+    struct AABB& getAABB() {
+        return aabb;
+    }
+    std::vector<struct MeshToTriangle>& getIntersected() {
+        return intersected;
+    }
+};
+
+class KDTree {
+    const int maxDepth = 3;
```

Comment:
Seems like a small maximum depth, if you haven't experiment with larger one

Reply:
Yeah, the improvement written in the improvements.txt file is achieved with a depth of 12.